### PR TITLE
Change early data flag to input file

### PR DESCRIPTION
**Pull request by xkqian**
## Description

Originally we have one input option to control whether enable early data or not. Currently we need one extra parameter to input various length of  early data. This pr will integrate them together.

This is part of #6542 

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [ ] **changelog** ~~provided, or~~ not required
- [ ] **backport** ~~done, or~~ not required
- [x] **tests** provided ~~, or not required~~

### CODE REVIEW

**Comment by tom-cosgrove-arm** [comment]
@xkqian Should this PR be part of the Early Data epic?

File: programs/ssl/ssl_client2.c [hunk 1]
```diff
@@ -543,7 +544,7 @@ struct options {
     int reproducible;           /* make communication reproducible          */
     int skip_close_notify;      /* skip sending the close_notify alert      */
 #if defined(MBEDTLS_SSL_EARLY_DATA)
-    int early_data;             /* support for early data                   */
+    const char *early_data_file; /* the file path of early data             */
```
Comment (ronald-cron-arm) [inline programs/ssl/ssl_client2.c:547]:
```suggestion
    const char *early_data_file; /* the path of the file containing the early data to send */
```

File: programs/ssl/ssl_client2.c [hunk 2]
```diff
@@ -347,8 +347,9 @@ int main(void)
 
 #if defined(MBEDTLS_SSL_EARLY_DATA)
 #define USAGE_EARLY_DATA \
-    "    early_data=%%d        default: 0 (disabled)\n"      \
-    "                        options: 0 (disabled), 1 (enabled)\n"
+    "    early_data=%%s      The file path to read early data from\n" \
+    "                         default: \"\" (do nothing)\n"                    \
+    "                         option: a file path\n"
```
Comment (ronald-cron-arm) [inline programs/ssl/ssl_client2.c:352]:
```suggestion
    "    early_data=%%s      The file path to read early data from\n" \
    "                        default: \"\" (do nothing)\n"            \
    "                        option: a file path\n"
```

File: programs/ssl/ssl_client2.c [hunk 3]
```diff
@@ -716,6 +717,46 @@ static int ssl_save_session_serialize(mbedtls_ssl_context *ssl,
     return ret;
 }
 
+#if defined(MBEDTLS_SSL_EARLY_DATA)
+
+#define MBEDTLS_ERR_EARLY_FILE_IO_ERROR -2
+
+int ssl_early_data_read_file(const char *path, unsigned char **buffer, size_t *length)
```
Comment (ronald-cron-arm) [inline programs/ssl/ssl_client2.c:724]:
```suggestion
static int ssl_read_file(const char *path, unsigned char **buffer, size_t *length)
```
This is a generic function to read the content of a file into a buffer, I do not think we should restrict that to early data.

File: programs/ssl/ssl_client2.c [hunk 4]
```diff
@@ -1971,7 +2008,16 @@ int main(int argc, char *argv[])
...
+    int early_data_enabled = 0;
+    size_t early_data_len;
+    if (strlen(opt.early_data_file) > 0 &&
+        ssl_early_data_read_file(opt.early_data_file,
+                                 &early_data, &early_data_len) == 0) {
+        early_data_enabled = MBEDTLS_SSL_EARLY_DATA_ENABLED;
+    } else {
+        early_data_enabled = MBEDTLS_SSL_EARLY_DATA_DISABLED;
+    }
+    mbedtls_ssl_conf_early_data(&conf, early_data_enabled);
```
Comment (ronald-cron-arm) [inline programs/ssl/ssl_client2.c:2020]:
```suggestion
    int early_data_enablement = MBEDTLS_SSL_EARLY_DATA_DISABLED;
    size_t early_data_len;
    if (strlen(opt.early_data_file) > 0 &&
        ssl_early_data_read_file(opt.early_data_file,
                                 &early_data, &early_data_len) == 0) {
        early_data_enablement = MBEDTLS_SSL_EARLY_DATA_ENABLED;
    }
    mbedtls_ssl_conf_early_data(&conf, early_data_enablement);
```

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
This looks almost good to me, just some minor suggestions.

File: programs/ssl/ssl_client2.c (hunk 3 above)
Comment (tom-cosgrove-arm) [reply programs/ssl/ssl_client2.c:724]: > a generic function to read the content of a file into a buffer

It's a generic function to read a text file into a buffer, and null-terminate that buffer. While it could be used to read an arbitrary (binary) file, maybe it would be best to call it `ssl_read_file_text()` (i.e. read text data from a file)

Comment (yuhaoth) [reply programs/ssl/ssl_client2.c:724]: I do not think we should wrap it like this. It looks too complex and has `malloc` fail risk. 

```
FILE * fp=fopen(...)
....
n=fread(buf, 1, sizeof(buf), fp);
ssl_write_early_data(ssl,buf,n)) ;

exit:
   fclose(fp);
```

Comment (xkqian) [reply programs/ssl/ssl_client2.c:724]: Rename the function to ssl_read_file_text() done and for the time being, keep it be protected by macro guard MBEDTLS_SSL_EARLY_DATA.

File: programs/ssl/ssl_client2.c [hunk 5]
```diff
@@ -347,8 +347,9 @@ int main(void)
 
 #if defined(MBEDTLS_SSL_EARLY_DATA)
 #define USAGE_EARLY_DATA \
-    "    early_data=%%d        default: 0 (disabled)\n"      \
-    "                        options: 0 (disabled), 1 (enabled)\n"
+    "    early_data=%%s      The file path to read early data from\n" \
```
Comment (yuhaoth) [inline programs/ssl/ssl_client2.c:350]: Then parameter name should align with internal variable name.

File: programs/ssl/ssl_client2.c [hunk 6]
```diff
@@ -716,6 +717,46 @@ static int ssl_save_session_serialize(mbedtls_ssl_context *ssl,
...
+    }
+
+    if (fread(*buffer, 1, *length, f) != *length) {
+        fclose(f);
+        return MBEDTLS_ERR_EARLY_FILE_IO_ERROR;
+    }
+
+    fclose(f);
+
+    (*buffer)[*length] = '\0';
```
Comment (yuhaoth) [inline programs/ssl/ssl_client2.c:754]:
```suggestion
```

File: programs/ssl/ssl_client2.c (hunk 4 above)
Comment (yuhaoth) [inline programs/ssl/ssl_client2.c:2020]: See also above suggestion.
```suggestion
    mbedtls_ssl_conf_early_data(&conf, early_data_fp == NULL ? MBETLS_SSL_EARLY_DATA_DISABLED : MBETLS_SSL_EARLY_DATA_ENABLED );
```

File: programs/ssl/ssl_client2.c [hunk 7]
```diff
@@ -3029,6 +3075,13 @@ int main(int argc, char *argv[])
     mbedtls_ssl_config_free(&conf);
     mbedtls_ssl_session_free(&saved_session);
 
+#if defined(MBEDTLS_SSL_EARLY_DATA)
+    if (early_data != NULL) {
+        mbedtls_platform_zeroize(early_data, early_data_len);
+    }
+    mbedtls_free(early_data);
```
Comment (yuhaoth) [inline programs/ssl/ssl_client2.c:3082]:
```suggestion
    if (early_data != NULL) {
        mbedtls_free(early_data);
    }
```
or 

```suggestion
    if (early_data_fp != NULL) {
        fclose(early_data_fp);
    }

```

**Review by yuhaoth — CHANGES_REQUESTED** [review]
Some comments. I think the implementation looks too complex and miss some features for tests.

File: programs/ssl/ssl_client2.c (hunk 5 above)
Comment (xkqian) [reply programs/ssl/ssl_client2.c:350]: Keep the input option as this name, just to align with the name from gnutls and openssl.

File: programs/ssl/ssl_client2.c (hunk 3 above)
Comment (xkqian) [reply programs/ssl/ssl_client2.c]: > I do not think we should wrap it like this. It looks too complex and has `malloc` fail risk.
> 
> ```
> FILE * fp=fopen(...)
> ....
> n=fread(buf, 1, sizeof(buf), fp);
> ssl_write_early_data(ssl,buf,n)) ;
> 
> exit:
>    fclose(fp);
> ```

If malloc failed, we think the test case should fail and stop, shouldn't it?  if so, we can avoid `malloc fail risk`.   And, cause we didn't know what size of the input file, will it be a better way to put `buf` on the stack?

File: programs/ssl/ssl_client2.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (xkqian) [reply programs/ssl/ssl_client2.c]: > Should we keep `early_data` parameter?
> How about the behaviors when early data is enable and `mbedtls_ssl_write_early_data()` is NOT called?Should/does client send `early_data` extension in this case?


--------------------
In my opinion, if we don't provide one valid input early data file, we should not enable early data.  That means:  in this design, there is no such way to enable early data but don't call mbedtls_ssl_write_early_data().  Anyway, we need talk about whether the design is reasonable and get one final decision. @ronald-cron-arm @yuhaoth @tom-cosgrove-arm

File: programs/ssl/ssl_client2.c (hunk 4 above)
Comment (xkqian) [reply programs/ssl/ssl_client2.c]: Simplify the file read and change it, a little different from this.

File: programs/ssl/ssl_client2.c (hunk 6 above)
Comment (xkqian) [reply programs/ssl/ssl_client2.c:756]: Removed.

File: programs/ssl/ssl_client2.c (hunk 3 above)
Comment (tom-cosgrove-arm) [reply programs/ssl/ssl_client2.c]: If `malloc()` is going to fail, either the file is huge or we have real problems. This is a program used by tests, not a part of the library, so let's keep using `malloc()`.

Comment (yuhaoth) [reply programs/ssl/ssl_client2.c]: `max_early_data_size` is up to `UINT_MAX` . It might cause `malloc()`.

File: programs/ssl/ssl_client2.c (hunk 5 above)
Comment (yuhaoth) [reply programs/ssl/ssl_client2.c:350]: prefer change it to `early_data_file` , same with `opt.early_data_file`

File: programs/ssl/ssl_client2.c (hunk 3 above)
Comment (tom-cosgrove-arm) [reply programs/ssl/ssl_client2.c]: > `max_early_data_size` is up to `UINT_MAX`

But that is too much for the stack also!

File: programs/ssl/ssl_client2.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (yuhaoth) [reply programs/ssl/ssl_client2.c]: >if we don't provide one valid input early data file, we should not enable early data. That means: in this design, there is no such way to enable early data but don't call mbedtls_ssl_write_early_data().

That's design of `ssl_server2`. I just want to make sure the behavior of above case and should we cover it in `ssl_client2`

File: programs/ssl/ssl_client2.c (hunk 3 above)
Comment (ronald-cron-arm) [reply programs/ssl/ssl_client2.c]: I think the plan is rather to read the file in chunks rather than in one go and store it in an allocated buffer. I think this is more aligned with real life thus I am rather in favor of the read by chunks option.

File: programs/ssl/ssl_client2.c [hunk 8]
```diff
@@ -1971,7 +1965,21 @@ int main(int argc, char *argv[])
...
+    if (strlen(opt.early_data_file) > 0) {
+        if ((early_data_fp = fopen(opt.early_data_file, "rb")) == NULL) {
+            mbedtls_printf("failed\n  ! Cannot open '%s' for reading.\n",
+                           opt.early_data_file);
+            goto exit;
+        }
+        early_data_len = fread(buf, 1, sizeof(buf), early_data_fp);
+        if (early_data_len > 0) {
+            early_data_enabled = MBEDTLS_SSL_EARLY_DATA_ENABLED;
+        }
```
Comment (yuhaoth) [inline programs/ssl/ssl_client2.c:1980]: Then we can cover enabled without zero length data
```suggestion
```

File: programs/ssl/ssl_client2.c [hunk 9]
```diff
@@ -1971,7 +1965,21 @@ int main(int argc, char *argv[])
...
+    int early_data_enabled = MBEDTLS_SSL_EARLY_DATA_DISABLED;
+    FILE *early_data_fp = NULL;
+    size_t early_data_len = 0;
+    if (strlen(opt.early_data_file) > 0) {
+        if ((early_data_fp = fopen(opt.early_data_file, "rb")) == NULL) {
+            mbedtls_printf("failed\n  ! Cannot open '%s' for reading.\n",
+                           opt.early_data_file);
+            goto exit;
+        }
+        early_data_len = fread(buf, 1, sizeof(buf), early_data_fp);
```
Comment (yuhaoth) [inline programs/ssl/ssl_client2.c:1977]: Nitpick comment, not necessary for this PR
```suggestion
```

File: programs/ssl/ssl_client2.c (hunk 5 above)
Comment (xkqian) [reply programs/ssl/ssl_client2.c:350]: See the option of GNUTLS, `--earlydata`, and the option of OpenSSL, `"-early_data"`. Both of them expect one input file but didn't name it as early_data_file. Why not keep aligned with them?
We will provide the comment to expect one input file, and internally stored it into one variable named xxx_file to facilitate the maintaining.

Comment (yuhaoth) [reply programs/ssl/ssl_client2.c:350]: Or change `opt.early_data_file` to `opt.early_data` , up to you. But that should be same

File: programs/ssl/ssl_client2.c (hunk 9 above)
Comment (xkqian) [reply programs/ssl/ssl_client2.c:1977]: Actually, this is related with the design we talked about in https://github.com/Mbed-TLS/mbedtls/pull/8596#discussion_r1416618029
We want a input file which isn't empty.
Anyway, it depend on the result of the discussion.

File: programs/ssl/ssl_client2.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply programs/ssl/ssl_client2.c]: > In my opinion, if we don't provide one valid input early data file, we should not enable early data. That means: in this design, there is no such way to enable early data but don't call mbedtls_ssl_write_early_data(). Anyway, we need talk about whether the design is reasonable and get one final decision. @ronald-cron-arm @yuhaoth @tom-cosgrove-arm

I think we need to test the edge but valid case where the client claims that it is going to send early data but it actually does not send any.

If I understand correctly, two possible approaches are proposed here for the early data ssl_client2 option(s):
1. only one  ssl_client2 option, early_data_file: if the option is not set early data is not enabled, if the option is set but the file is empty, early data is enabled but no early data send ...
2. two ssl_client2 options, one to enable and the other to pass the path of the file containing the data to transfer. In that case no file or an empty file are equivalent.

Both work but the second one is better practice I would say (but I can be wrong), separation of purpose.

File: programs/ssl/ssl_client2.c (hunk 3 above)
Comment (xkqian) [reply programs/ssl/ssl_client2.c]: Have revert the one go reading, and instead, will read it in chunks.

File: programs/ssl/ssl_client2.c [hunk 10]
```diff
@@ -957,7 +914,7 @@ int main(int argc, char *argv[])
     opt.groups              = DFL_GROUPS;
     opt.sig_algs            = DFL_SIG_ALGS;
 #if defined(MBEDTLS_SSL_EARLY_DATA)
-    opt.early_data_file     = DFL_EARLY_DATA_FILE;
+    opt.early_data          = DFL_EARLY_DATA_FILE;
```
Comment (ronald-cron-arm) [inline programs/ssl/ssl_client2.c:917]:
```suggestion
    opt.early_data          = DFL_EARLY_DATA;
```

File: programs/ssl/ssl_client2.c [hunk 11]
```diff
@@ -52,7 +52,7 @@ int main(void)
 #define DFL_KEY_OPAQUE          0
 #define DFL_KEY_PWD             ""
 #define DFL_PSK                 ""
-#define DFL_EARLY_DATA          MBEDTLS_SSL_EARLY_DATA_DISABLED
+#define DFL_EARLY_DATA_FILE     ""
```
Comment (ronald-cron-arm) [inline programs/ssl/ssl_client2.c:55]:
```suggestion
#define DFL_EARLY_DATA          ""
```

File: programs/ssl/ssl_client2.c [hunk 12]
```diff
@@ -543,7 +544,8 @@ struct options {
     int reproducible;           /* make communication reproducible          */
     int skip_close_notify;      /* skip sending the close_notify alert      */
 #if defined(MBEDTLS_SSL_EARLY_DATA)
-    int early_data;             /* support for early data                   */
+    const char *early_data;     /* the path of the file containing the
+                                 * early data to send                       */
```
Comment (ronald-cron-arm) [inline programs/ssl/ssl_client2.c:548]:
```suggestion
    const char *early_data;     /* the path of the file to read early data from */
```

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
Some minor comments

File: programs/ssl/ssl_client2.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (yuhaoth) [reply programs/ssl/ssl_client2.c]: I prefer second one with some modifications. 
- `early_data` : reserve, enable/disable early data
- `early_data_file`: path of early data file. When set, `mbedtls_ssl_write_handshake()` should be called in next PR. Otherwise, `mbedtls_ssl_handshake()` will be called. 
In next PR, we can make final decision which one should be tested. 

This PR is for adding options. I think we should not limit the possible cases in this PR.

[the remark this answers is not in the document]
Comment (xkqian) [reply programs/ssl/ssl_client2.c]: Thank you for your suggestions, will follow up the second way.

`2 .two ssl_client2 options, one to enable and the other to pass the path of the file containing the data to transfer. In that case no file or an empty file are equivalent.`

File: programs/ssl/ssl_client2.c (hunk 10 above)
Comment (xkqian) [reply programs/ssl/ssl_client2.c]: Provide two options, both DFL_EARLY_DATA_FILE and DFL_EARLY_DATA are exist, please help review again.

File: programs/ssl/ssl_client2.c (hunk 11 above)
Comment (xkqian) [reply programs/ssl/ssl_client2.c:56]: ditto.

File: programs/ssl/ssl_client2.c [hunk 13]
```diff
@@ -1205,6 +1215,8 @@ int main(int argc, char *argv[])
                     break;
                 default: goto usage;
             }
+        } else if (strcmp(p, "early_data_file") == 0) {
+            opt.early_data_file = q;
```
Comment (yuhaoth) [inline programs/ssl/ssl_client2.c:1219]: Nitpick comments. move fopen here
```suggestion
        } else if (strcmp(p, "early_data_file") == 0) {
            opt.early_data_file = q;
            if ((early_data_fp = fopen(opt.early_data_file, "rb")) == NULL) {
                mbedtls_printf("failed\n  ! Cannot open '%s' for reading.\n",
                               opt.early_data_file);
                ret = MBEDTLS_ERR_SSL_INTERNAL_ERROR;
                goto exit;
            }

```

File: programs/ssl/ssl_client2.c [hunk 14]
```diff
@@ -2995,6 +3007,25 @@ int main(int argc, char *argv[])
                            (unsigned int) -ret);
             goto exit;
         }
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3) && defined(MBEDTLS_SSL_EARLY_DATA)
```
Comment (yuhaoth) [inline programs/ssl/ssl_client2.c:3010]:
```suggestion
#if defined(MBEDTLS_SSL_EARLY_DATA)
```

File: programs/ssl/ssl_client2.c [hunk 15]
```diff
@@ -2995,6 +3007,25 @@ int main(int argc, char *argv[])
...
+        if (opt.early_data == MBEDTLS_SSL_EARLY_DATA_ENABLED
+            && strlen(opt.early_data_file) > 0) {
```
Comment (yuhaoth) [inline programs/ssl/ssl_client2.c:3012]: I do not think this limitation match 2nd way from Ronald
`2 .two ssl_client2 options, one to enable and the other to pass the path of the file containing the data to transfer. In that case no file or an empty file are equivalent.`

As my understand, 
- It does not means we should not open early data file when disable. 
- `empty file` does not means file name len is zero. You should check the file length. 

On my view, this check make things complex.

**Review by yuhaoth — CHANGES_REQUESTED** [review]
Thanks for address my comments. 

Nitpick comments are base on my suggestion.  
And I do not think the latest version match the suggestion from Ronald.

File: programs/ssl/ssl_client2.c (hunk 15 above)
Comment (ronald-cron-arm) [reply programs/ssl/ssl_client2.c:3012]: If the two options are at their default value then better to not do anything I'd say.
Then if early_data == MBEDTLS_SSL_EARLY_DATA_ENABLED call write_early_data in all cases (with a 0 length if no file defined or an empty file.
If early_data == MBEDTLS_SSL_EARLY_DATA_DISABLED call also write_early_data in all cases (with a 0 length if the file is empty) and expect the function to return in error.

**Comment by ronald-cron-arm** [comment]
I am fine with the two options solution. I put a comment in one of Jerry's suggestion.

File: programs/ssl/ssl_client2.c (hunk 15 above)
Comment (yuhaoth) [reply programs/ssl/ssl_client2.c]: understand.

**Review by yuhaoth — APPROVED** [review]
LGTM

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
While thinking about what option to add to ssl_client2 for early data testing I think we have lost (at least that's my case) sight of its scope. It is mainly about testing the various TLS options and the compatibility with GnuTLS and OpenSSL. Negative and edge cases testing is not really in scope.

Thus I'd say let's keep things simple, for the time being at least:
one option early_data_file whose default value is "". If a valid file path is set to early_data_file, we enable early data on the client and in the reconnection case, before to call `mbedtls_ssl_handshake()` to reconnect we call `mbedtls_ssl_write_early_data()` that tries to write the `early_data_file` file data.

When we add more testing for early data we will see if they best fit in ssl-opt.sh or test_suite_ssl and we need to add other options in ssl_client2.

**Comment by yuhaoth** [comment]
> While thinking about what option to add to ssl_client2 for early data testing I think we have lost (at least that's my case) sight of its scope. It is mainly about testing the various TLS options and the compatibility with GnuTLS and OpenSSL. Negative and edge cases testing is not really in scope.
> 
> Thus I'd say let's keep things simple, for the time being at least: one option early_data_file whose default value is "". If a valid file path is set to early_data_file, we enable early data on the client and in the reconnection case, before to call `mbedtls_ssl_handshake()` to reconnect we call `mbedtls_ssl_write_early_data()` that tries to write the `early_data_file` file data.
> 

My target is to simple this PR. Let's make final decision in next PRs . 
But if only one option, we have to do more things in this PR.  `early_data` option come from `development`, it must be removed.  I do not thinks we should remove or change it in this PR. That make things complex. 

For the two options from your previous comments, I do not think empty file and invalid file are same.  If they are same, we have to check the file length before call `mbedtls_ssl_{handshake, write_early_data}()` and reduce possible test case. And I do not mean we MUST add the test in `ssl-opt.sh`. But the amount of code is less and the test coverage is more, why not do like that?


> When we add more testing for early data we will see if they best fit in ssl-opt.sh or test_suite_ssl and we need to add other options in ssl_client2.

Agree.  
I think this PR just do two things enough.
- add `early_data_file` option
- add a empty function for writing early data in next PRs .

**Comment by xkqian** [comment]
I just revert the code to `use the one option`, which uses `early_data` as the input early data file. 
In this PR, we can only replace the option, and in  next PR, we can add the test code including:  ssl_write_early_data(), and how to call it.
I think we can review it base one current version.

**Review by ronald-cron-arm — DISMISSED** [review]
If with this changes you have what you need for your coming PRs implementing `mbedtls_ssl_write_early_data()` this is good enough to me.

File: programs/ssl/ssl_client2.c [hunk 16]
```diff
@@ -1971,7 +1964,17 @@ int main(int argc, char *argv[])
...
+    int early_data_enabled = MBEDTLS_SSL_EARLY_DATA_DISABLED;
+    FILE *early_data_fp = NULL;
+    if (strlen(opt.early_data) > 0) {
+        if ((early_data_fp = fopen(opt.early_data, "rb")) == NULL) {
+            mbedtls_printf("failed\n  ! Cannot open '%s' for reading.\n",
+                           opt.early_data);
+            goto exit;
+        }
+        early_data_enabled = MBEDTLS_SSL_EARLY_DATA_ENABLED;
+    }
```
Comment (yuhaoth) [inline programs/ssl/ssl_client2.c:1976]: Nitpick comment and move it to option check like previous version
```suggestion
       mbedtls_ssl_conf_early_data(&conf, early_data_fp ?MBEDTLS_SSL_EARLY_DATA_ENABLE: MBEDTLS_SSL_EARLY_DATA_DISABLE);
```

**Review by yuhaoth — COMMENTED** [review]
It has got CI failure
One change request and one nitpick comments.

File: programs/ssl/ssl_client2.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (xkqian) [reply programs/ssl/ssl_client2.c:1976]: The CI failure is caused by the uninitialized file pointer early_data_fp.
opt.early_data will be default to "", this check maybe useless...

Comment (xkqian) [reply programs/ssl/ssl_client2.c:1976]: I just see there are file processings like `context_file`, will treat it like what I have done, so I prefer that way.

**Review by ronald-cron-arm — APPROVED** [review]

**Comment by tom-cosgrove-arm** [comment]
@yuhaoth are you happy with this now, too?

File: programs/ssl/ssl_client2.c [hunk 17]
```diff
@@ -1971,7 +1968,16 @@ int main(int argc, char *argv[])
...
+    int early_data_enabled = MBEDTLS_SSL_EARLY_DATA_DISABLED;
+    if (strlen(opt.early_data) > 0) {
+        if ((early_data_fp = fopen(opt.early_data, "rb")) == NULL) {
+            mbedtls_printf("failed\n  ! Cannot open '%s' for reading.\n",
+                           opt.early_data);
+            goto exit;
+        }
+        early_data_enabled = MBEDTLS_SSL_EARLY_DATA_ENABLED;
+    }
+    mbedtls_ssl_conf_early_data(&conf, early_data_enabled);
```
Comment (tom-cosgrove-arm) [inline programs/ssl/ssl_client2.c:1980]: (nit) presumably early data defaults to disabled, so we could move this into the `if()` and remove the variable?
